### DISCUSSION TITLE
Question about ensembling and 5-fold cross-validation

### DISCUSSION THREAD

**Discussion by m-aminiz** [MIC-DKFZ/nnUNet#824] [General]
Hi Fabian
unfortunately, I didn't understand 5-fold cross-validation well.
for the sake of  5-fold cross-validation I did this:
I ran below command just once:
!nnUNet_plan_and_preprocess -t 1
and after that I ran training command 5 time for creating each fold's mode like this:
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 0 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 1 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 2 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 3 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 4 --npz
Is this correct?
should I run "nnUNet_plan_and_preprocess" before every training command?

**Comment by FabianIsensee** [comment]
You only need to run `nnUNet_plan_and_preprocess` once. The preprocessed data is saved in your nnUNet_preprocessed folder

> I ran training command 5 time for creating each fold's mode like this:
> !nnUNet_train 3d_fullres nnUNetTrainerV2 1 0 --npz
> !nnUNet_train 3d_fullres nnUNetTrainerV2 1 1 --npz
> !nnUNet_train 3d_fullres nnUNetTrainerV2 1 2 --npz
> !nnUNet_train 3d_fullres nnUNetTrainerV2 1 3 --npz
> !nnUNet_train 3d_fullres nnUNetTrainerV2 1 4 --npz

This is correct

**Reply by m-aminiz** [reply]
> You only need to run `nnUNet_plan_and_preprocess` once. The preprocessed data is saved in your nnUNet_preprocessed folder
> 
> > I ran training command 5 time for creating each fold's mode like this:
> > !nnUNet_train 3d_fullres nnUNetTrainerV2 1 0 --npz
> > !nnUNet_train 3d_fullres nnUNetTrainerV2 1 1 --npz
> > !nnUNet_train 3d_fullres nnUNetTrainerV2 1 2 --npz
> > !nnUNet_train 3d_fullres nnUNetTrainerV2 1 3 --npz
> > !nnUNet_train 3d_fullres nnUNetTrainerV2 1 4 --npz
> 
> This is correct

thank you
Which command should I use after I have completed all folds training?
If I use the below command, does it use all the folds model for predicting?
1) nnUNet_predict -i INPUT_FOLDER -o OUTPUT_FOLDER -t TASK_NAME_OR_ID -m CONFIGURATION --save_npz

Maybe I should use the following command?
2) nnUNet_ensemble -f FOLDER1 FOLDER2 ... -o OUTPUT_FOLDER -pp POSTPROCESSING_FILE
I think for this one FOLDER1 is the path of fold1 model, FOLDER2 is the path of fold2 model and so on

which one should I use?

**Reply by FabianIsensee** [reply]
`nnUNet_predict -i INPUT_FOLDER -o OUTPUT_FOLDER -t TASK_NAME_OR_ID -m CONFIGURATION --save_npz`

and I recommend you read the documentation a little more ;-)

**Comment by Jun-medicalAI** [comment]
I have a question here. 
We must run the following items Five times to finish the 5-fold? Is there one command to finish the 5-fold?

> !nnUNet_train 3d_fullres nnUNetTrainerV2 1 0 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 1 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 2 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 3 --npz
!nnUNet_train 3d_fullres nnUNetTrainerV2 1 4 --npz

**Comment by FabianIsensee** [comment]
Hi @Jun-electrophysiology , there is no one command to run all of the folds because it is best to run them at the same time, each on a separate GPU. If you want to run them all at once you can just paste these commands in a shell script and run that instead
Best,
Fabian
